Library/OcVariableRuntimeLib/OcVariableRuntimeLib.c

```diff
@@ -0,0 +1,621 @@
+LocateNvramDir (
+  OUT EFI_FILE_PROTOCOL  **NvramDir
+  )
+{
+  EFI_STATUS         Status;
+  EFI_FILE_PROTOCOL  *Root;
+
+  if ((mStorageContext == NULL) || (mNvramConfig == NULL)) {
+    return EFI_NOT_READY;
+  }
+
+  if (mStorageContext->FileSystem == NULL) {
+    return EFI_NOT_FOUND;
+  }
+
+  Status = mStorageContext->FileSystem->OpenVolume (mStorageContext->FileSystem, &Root);
+  if (EFI_ERROR (Status)) {
+    return EFI_NOT_FOUND;
+  }
+
+  Status = OcSafeFileOpen (
+             Root,
+             NvramDir,
+             OPEN_CORE_NVRAM_ROOT_PATH,
+             EFI_FILE_MODE_READ | EFI_FILE_MODE_WRITE | EFI_FILE_MODE_CREATE,
+             EFI_FILE_DIRECTORY
+             );
+
+  return Status;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+LoadNvram (
+  IN OC_STORAGE_CONTEXT  *StorageContext,
+  IN OC_NVRAM_CONFIG     *NvramConfig
+  )
+{
+  EFI_STATUS             Status;
+  EFI_FILE_PROTOCOL      *NvramDir;
+  UINT8                  *FileBuffer;
+  UINT32                 FileSize;
+  BOOLEAN                IsValid;
+  OC_NVRAM_STORAGE       NvramStorage;
+  UINT32                 GuidIndex;
+  UINT32                 VariableIndex;
+  GUID                   VariableGuid;
+  OC_ASSOC               *VariableMap;
+  OC_NVRAM_LEGACY_ENTRY  *SchemaEntry;
+
+  if ((mStorageContext != NULL) || (mNvramConfig != NULL)) {
+    return EFI_ALREADY_STARTED;
+  }
+
+  if ((StorageContext == NULL) || (NvramConfig == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  mStorageContext = StorageContext;
+  mNvramConfig    = NvramConfig;
+
+  Status = LocateNvramDir (&NvramDir);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  FileBuffer = OcReadFileFromDirectory (NvramDir, OPEN_CORE_NVRAM_FILENAME, &FileSize, NVRAM_PLIST_MAX_SIZE);
+  if (FileBuffer == NULL) {
+    FileBuffer = OcReadFileFromDirectory (NvramDir, OPEN_CORE_NVRAM_FALLBACK_FILENAME, &FileSize, NVRAM_PLIST_MAX_SIZE);
+  }
+
+  NvramDir->Close (NvramDir);
+  if (FileBuffer == NULL) {
+    return EFI_NOT_FOUND;
+  }
+
+  OC_NVRAM_STORAGE_CONSTRUCT (&NvramStorage, sizeof (NvramStorage));
+  IsValid = ParseSerialized (&NvramStorage, &mNvramStorageRootSchema, FileBuffer, FileSize, NULL);
+  FreePool (FileBuffer);
+
+  if (!IsValid) {
+    OC_NVRAM_STORAGE_DESTRUCT (&NvramStorage, sizeof (NvramStorage));
+    return EFI_UNSUPPORTED;
+  }
+
+  if (NvramStorage.Version != OC_NVRAM_STORAGE_VERSION) {
+    OC_NVRAM_STORAGE_DESTRUCT (&NvramStorage, sizeof (NvramStorage));
+    return EFI_UNSUPPORTED;
+  }
+
+  for (GuidIndex = 0; GuidIndex < NvramStorage.Add.Count; ++GuidIndex) {
+    Status = OcProcessVariableGuid (
+               OC_BLOB_GET (NvramStorage.Add.Keys[GuidIndex]),
+               &VariableGuid,
+               &mNvramConfig->Legacy,
+               &SchemaEntry
+               );
+
+    if (EFI_ERROR (Status)) {
+      continue;
+    }
+
+    VariableMap = NvramStorage.Add.Values[GuidIndex];
+
+    //
+    // TODO: Is OC LegacyOverwrite setting still needed, given that this code
```

Comment:
Hrm, why would it be empty?

Reply:
When we've initialised emulated NVRAM, before we anything into it from our file, it must be empty, right?  But I guess not - things in the emulated NVRAM driver itself (in the case of OC-only use) and in the rest of OpenDuet (in the case of OpenDuet) could already have populated certain things, right?

Reply:
(Noting that in the case of OC previously, there was another use entirely, which was when reading from `nvram.plist`, but writing over non-emulated, normal NVRAM. But this use no longer exists.)

Reply:
I'm assuming that my new comment added here (basically covering my second comment above) captures why this is still useful.